File: Surface/Surface/SfRect.m
//
//  SfRect.m
//  Surface
//
//  Created by Webchimp on 18/03/16.
//  Copyright © 2016 Webchimp. All rights reserved.
//

#import "SfRect.h"

@implementation SfRect

- (id) init
{
    self = [super init];
    if (self) {
        
        self.x = 0;
        self.y = 0;
        self.width = 0;
        self.height = 0;
        
    }
    return self;
}

- (void)setRectEmpty {
    
    self.x = 0;
    self.y = 0;
    self.width = 0;
    self.height = 0;
}

- (void)setRect:(float)x Y:(float)y Width:(float)width Height:(float)height {
    
    self.x = x;
    self.y = y;
    self.width = width;
    self.height = height;
}

- (void)moveRect:(float)dx DY:(float)dy {
    
    self.x += dx;
    self.y += dy;
}

- (void)offsetRect:(float)dw DH:(float)dh {
    
    self.width += dw;
    self.height += dh;
}

- (BOOL)isEmpty {
    
    if ( !self.x && !self.y && !self.width && !self.height ){
        return true;
    }
    return false;
}

- (void)unionRect:(SfRect *)rect {
    
    if ( rect != nil ) {
        
        float x2 = rect.x + rect.width;
        float y2 = rect.y + rect.height;
        
        self.width = self.x - x2;
        self.height = self.y - y2;
    }
}

- (void)intersectRect:(SfRect *)rect {
    
    if ( rect != nil ) {
        
        float x = rect.x;
        float y = rect.y;
        
        float x2 = (self.x + self.width) - x;
        float y2 = (self.y + self.height) - y;
        
        [self setRect:x Y:y Width:x2 Height:y2];
    }
    
}

- (void)subtractRect:(SfRect *)rect {
    
    if ( rect != nil ) {
        
        self.width = rect.x - self.x;
        self.height = rect.y - self.y;
    }
}

- (BOOL)pointInRect:(SfPoint *)point {
    
    BOOL ret = false;
    
    if ( point != nil ) {
        ret = (point.x >= self.x && point.y >= self.y && point.x <= (self.x + self.width) && point.y <= (self.y + self.height));
    }
    return ret;

}



@end

File: Swift/V2/SfRect.swift
//
//  SfRect.swift
//  Surface Swift
//
//  Created by Webchimp on 26/09/16.
//  Copyright © 2016 AppBuilders. All rights reserved.
//

import Foundation

class SfRect: NSObject {
    
    var x: Float! = 0.0;
    var y: Float! = 0.0;
    var width: Float! = 0.0;
    var height: Float! = 0.0;
    
    func setRectEmpty() -> Void {
        
        self.x = 0;
        self.y = 0;
        self.width = 0;
        self.height = 0;
    }
    
    func setRect(x:Float, y:Float, width:Float, height:Float) -> Void {
        
        self.x = x;
        self.y = y;
        self.width = width;
        self.height = height;
    }
    
    func moveRect(dx:Float, dy:Float) -> Void {
        
        self.x = self.x + dx;
        self.y = self.y + dy;
    }
    
    func offsetRect(dw:Float, dh:Float) ->  Void {
        
        self.width = dw;
        self.height = dh;
    }
    
    func isEmpty() -> Bool {
        
        if ( self.x == nil && self.y == nil && self.width == nil && self.height == nil ) {
            return true;
        }
        
        return false;
    }
    
    func unionRect(rect:SfRect!) {
        
        if ( rect != nil ) {
            
            let x2:Float! = rect.x + rect.width;
            let y2:Float! = rect.y + rect.height;
            
            self.width = self.x - x2;
            self.height = self.y - y2;
            
        }
    }
    
    func intersectRect(rect:SfRect!) {
        
        if ( rect != nil ) {
            
            let x:Float! = rect.x;
            let y:Float! = rect.y;
            
            let x2:Float! = ( self.x + self.width ) - x;
            let y2:Float! = ( self.y + self.height ) - y;
            
            self.setRect(x, y: y, width: x2, height: y2);
        }
    }
    
    func subtractRect(rect:SfRect!) {
        
        if ( rect != nil ) {
            
            self.width = rect.x - self.x;
            self.height = rect.y - self.y;
        }
    }
    
    func pointInRect(point:SfPoint!) -> Bool {
        
        var ret:Bool! = false;
        
        if ( point != nil ) {
            ret = ( point.x >= self.x && point.y >= self.y && point.x <= (self.x + self.width) && point.y <= (self.y + self.height) );
        }
        return true;
    }

}
